threescale/src/structs.rs

```diff
@@ -45,11 +45,168 @@ pub struct UsageReport {
     pub max_value: u64,
 }
 
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct AppId(String);
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct AppKey(String);
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct UserKey(String);
+
+impl AsRef<str> for AppId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl AsRef<str> for AppKey {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl AsRef<str> for UserKey {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl From<&str> for UserKey {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for AppId {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for AppKey {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ServiceToken(String);
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ServiceId(String);
+
+impl AsRef<str> for ServiceToken {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl AsRef<str> for ServiceId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl From<&str> for ServiceToken {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for ServiceId {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CacheKey(ServiceId, AppIdentifier);
+
+impl<'a> CacheKey {
+    pub fn as_string(&self) -> String {
+        format!("{}_{}", self.0 .0, self.1.as_string())
+    }
+
+    pub fn default() -> CacheKey {
+        CacheKey(
+            ServiceId(String::new()),
+            AppIdentifier::UserKey(UserKey(String::new())),
+        )
+    }
+
+    pub fn service_id(&'a self) -> &'a ServiceId {
+        &self.0
+    }
+
+    pub fn app_id(&'a self) -> &'a AppIdentifier {
+        &self.1
+    }
+
+    pub fn from(a: &ServiceId, b: &AppIdentifier) -> CacheKey {
+        CacheKey {
+            0: a.clone(),
+            1: b.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum AppIdentifier {
+    AppId(AppId, Option<AppKey>),
+    UserKey(UserKey),
+}
+
+impl From<AppId> for AppIdentifier {
+    fn from(a: AppId) -> Self {
+        AppIdentifier::AppId(a, None)
+    }
+}
+
+impl From<(AppId, AppKey)> for AppIdentifier {
+    fn from(a: (AppId, AppKey)) -> Self {
+        AppIdentifier::AppId(a.0, Some(a.1))
+    }
+}
+
+impl From<UserKey> for AppIdentifier {
+    fn from(u: UserKey) -> Self {
+        AppIdentifier::UserKey(u)
+    }
+}
+
+impl AppIdentifier {
+    // This cannot return a reference because app_id:app_key needs to be generated
+    pub fn as_string(&self) -> String {
```

Comment:
Can you return the type of `AppIdentifier` as well ? Because when consuming these data and creating threescalers application , we need to know which type of key we have.